# Python for Tableau Intro:

## Imports

In [47]:
# Bring in all necessary imports for notebook
import pandas as pd
import plotly.express as px

## Step 1: Gather the data

This data was gathered from a Kaggle Competition located [here](https://www.kaggle.com/competitions/titanic/data?select=train.csv)
We will build a wrangle function to bring in the data as well as clean it!

## Step 2: Transform and Clean
This wrangle function will transform and clean all data for us as it reads in the data.

In [48]:
# Create wrangle function
def wrangle(filepath):
    # Pre-exploration: only returning a dataframe object
    return pd.read_csv(filepath)

def wrangle(filepath):
    # Post exploration: Any necessary edits
    df = pd.read_csv(filepath)
    
    # Change the fare column to be rounded to two decimal places:
    df['Fare'] = round(df['Fare'], 2)
    
    # Return dataframe object
    return df

In [49]:
# Use the function to create a dataframe
df = wrangle(r'C:\Users\Alex Lucchesi\OneDrive\Desktop\tableau_teaching_day_1\train.csv')

# View the first 5 rows
df.head()

PassengerId  Survived  Pclass   
0            1         0       3  \
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp   
0                            Braund, Mr. Owen Harris    male  22.0      1  \
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket   Fare Cabin Embarked  
0      0         A/5 21171   7.25   NaN        S  
1      0          PC 17599  71.28   C85        C  
2      0  STON/O2. 3101282   7.92   NaN        S  
3      0            113803  53.10  C123        S  
4      0            373450   8.05   NaN        S

### What do we notice?
In other words, what do we need to fix? Start building a list of items we see and the reason we want to change any of it.


- Parch: Column (drop?)
- Fare: Change to two decimal places
- Ensure survived is a binary (boolean) column
- Age: Ensure age is an integer, not a float
- Change column names to follow proper naming conventions
- Look over embarked values to ensure that the naming conventions follow a good vizualization process.
- Look over ticket column for values and uniques, see if any changes are needed

## Step 2: Exploration: Also called Exploratory Data Analysis

In this section, we will focus on exploring the data and cycling back to make any necessary changes to the data itself using tranformation and tidying.

To begin with, we will explore the Parch column:


We asked the question: Should we drop this column? Looks like it is filled with 0's.

We will start with checking unique values in the column:

In [50]:
# Using pandas functionality, check the unique values of the Parch column
print(df['Parch'].unique())

# Using pandas, check the number of unique values
print(len(df))
print(df['Parch'].value_counts())
print(df['Parch'].value_counts(normalize=True)*100)

# Calculate a percentage of it:
print((678/891)*100)

[0 1 2 5 3 4 6]
891
Parch
0    678
1    118
2     80
5      5
3      5
4      4
6      1
Name: count, dtype: int64
Parch
0    76.094276
1    13.243547
2     8.978676
5     0.561167
3     0.561167
4     0.448934
6     0.112233
Name: proportion, dtype: float64
76.0942760942761


This column represents each parent/child that each passenger had aboard the ship. While the majority class of the column is 0, it is still an effective communication and should be left present within the data source. It does not need to be used for modeling and can be dropped by my ML Engineers in their testing split.

### Question 2:
Edit the Fare column to be rounded to two decimal places.

In [51]:
# To do so, we will use the built-in round function:
df['Fare'] = round(df['Fare'], 2)

# View the data
df.head()

PassengerId  Survived  Pclass   
0            1         0       3  \
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp   
0                            Braund, Mr. Owen Harris    male  22.0      1  \
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket   Fare Cabin Embarked  
0      0         A/5 21171   7.25   NaN        S  
1      0          PC 17599  71.28   C85        C  
2      0  STON/O2. 3101282   7.92   NaN        S  
3      0            113803  53.10  C123        S  
4      0            373450   8.05   NaN        S

We use the built in Python Round function to round the data, then we will add this into our wrangle function above, once we have functionality established.

### Question 3:
Is the Survived column a binary column?

In [52]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [53]:
# What integer values are present?
df['Survived'].value_counts()

Survived
0    549
1    342
Name: count, dtype: int64

While the datatype itself is listed as an integer type, an overview of the survived column has gleaned us the insight that it is filled with 0, 1. This means that the column is a boolean that was converted into integers. We will leave it the same for modeling purposes. Can convert type if needed.

### Question 4:
How do we convert age from an integer value to a float value?

To start with, we should look at the:

- Unique values
- Number of unique values
- Are there any Null values present
- And the count of any null values

In [54]:
# How many unique values?
print(f'Number of unique values: {df["Age"].nunique()}')

# What are those values?
print(df['Age'].value_counts())

# Last thing is to check for null values:
# Filtering to the null values
print(df['Age'][df['Age'].isna() == True])

# Adding up the number of null values in a Series
print(df['Age'].isna().sum())

# First thing first, we would need to deal with these null values:
# Fill them with a zero
df['Age'].fillna(0.00, inplace=True)
print(df['Age'].isna().sum())
# We can determine that we can absolutely make this change:
df['Age'] = df['Age'].astype(int)

# View the change
df.head()

Number of unique values: 88
Age
24.00    30
22.00    27
18.00    26
19.00    25
28.00    25
         ..
36.50     1
55.50     1
0.92      1
23.50     1
74.00     1
Name: count, Length: 88, dtype: int64
5     NaN
17    NaN
19    NaN
26    NaN
28    NaN
       ..
859   NaN
863   NaN
868   NaN
878   NaN
888   NaN
Name: Age, Length: 177, dtype: float64
177
0


PassengerId  Survived  Pclass   
0            1         0       3  \
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex  Age  SibSp   
0                            Braund, Mr. Owen Harris    male   22      1  \
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   38      1   
2                             Heikkinen, Miss. Laina  female   26      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   35      1   
4                           Allen, Mr. William Henry    male   35      0   

   Parch            Ticket   Fare Cabin Embarked  
0      0         A/5 21171   7.25   NaN        S  
1      0          PC 17599  71.28   C85        C  
2      0  STON/O2. 3101282   7.92   NaN        S  
3      0            113803  53.10  C123        S  
4      0            373450   8.05   NaN        S

By exploring the age column, we came to the following conclusion:
- The values for age should not be a float. A float did not provide any additional useful information. Instead, we transformed it into an integer
- There were 177 missing values that needed to be handled before we began changing the data over.

### Question 5:
Change the column names to match snake casing:

In [55]:
# First, we will need to find a way to see if it is two words combined into 1:
col = list(df.columns)
col
# We can see that this is a huge task to try and undertake. It may be easier to use rename function instead:
df.rename(columns={
    'PassengerId': 'passenger_id',
    'Survived':'survived',
    'Pclass': 'p_class',
    'Name': 'name',
    'Sex': 'gender',
    'Age': 'age',
    'SibSp': 'siblings_or_spouse_aboard',
    'Parch': "parents_or_child_aboard",
    'Ticket':'ticket',
    'Fare': 'fare',
    'Cabin': 'cabin',
    'Embarked':'embarked'
}, inplace=True)

df.head()


passenger_id  survived  p_class   
0             1         0        3  \
1             2         1        1   
2             3         1        3   
3             4         1        1   
4             5         0        3   

                                                name  gender  age   
0                            Braund, Mr. Owen Harris    male   22  \
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   38   
2                             Heikkinen, Miss. Laina  female   26   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   35   
4                           Allen, Mr. William Henry    male   35   

   siblings_or_spouse_aboard  parents_or_child_aboard            ticket   
0                          1                        0         A/5 21171  \
1                          1                        0          PC 17599   
2                          0                        0  STON/O2. 3101282   
3                          1                        0            113803   
4                          0                        0            373450   

    fare cabin embarked  
0   7.25   NaN        S  
1  71.28   C85        C  
2   7.92   NaN        S  
3  53.10  C123        S  
4   8.05   NaN        S

### Question 6: 
Ensure that the values for the embarked column match what we need for visualizations:

First thing, we will create a visualization of the embarked column and try to glean understanding from it for an explanation of results

In [56]:
px.bar(df, x='embarked')

The analysis of this question has led us to the answer of it depends. If the situation required a change in the values for this column in order to communicate understanding, we can easily do so. Otherwise, the information shared, is the same as the information gained.

### Question 7:
Check the ticket column for any necessary changes we need to do to it:

In [57]:
df['ticket'].nunique()

681

Too many unique values to either visualize or model. We should drop this column

In [58]:
df = df.drop(columns=['ticket'])

df.head()

passenger_id  survived  p_class   
0             1         0        3  \
1             2         1        1   
2             3         1        3   
3             4         1        1   
4             5         0        3   

                                                name  gender  age   
0                            Braund, Mr. Owen Harris    male   22  \
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   38   
2                             Heikkinen, Miss. Laina  female   26   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   35   
4                           Allen, Mr. William Henry    male   35   

   siblings_or_spouse_aboard  parents_or_child_aboard   fare cabin embarked  
0                          1                        0   7.25   NaN        S  
1                          1                        0  71.28   C85        C  
2                          0                        0   7.92   NaN        S  
3                          1                        0  53.10  C123        S  
4                          0                        0   8.05   NaN        S